'Client.Endpoint()'  returns 'context canceled' while reading the Body

Hello! I have a question which IDK if it's a missus on my part and there is another way to do it.

I'm writing a client which consumes a MS that returns binary data. I'm using the `kithttp.BufferedStream(true)` while `NewClient` to make sure the body is not closed so I can read it when I want and not have it buffered.

This works fine with my implementation of the `decode*` that I use on the `NewClient` except if the body is longer than `3966` which then stops reading from the body.

This is my `decode*`:

```go
func decode(_ context.Context, r *http.Response) (interface{}, error) {
        pr, pw := io.Pipe()

        go func() {
                defer r.Body.Close()
                defer pw.Close()
                i, err := io.Copy(pw, r.Body)
                fmt.Println(i, err)
        }()
        // IOR is a io.Reader
        return response{IOR: pr}, nil
}
```

The intention is to only read from the `r.Body` when I need it, that's why I'm using a `io.Pipe`, and also because I have to `r.Body.Close()` when it ends (if not, just passing the `r.Body` to the `response.IOR` would work, but then the `r.Body` wold not closed, no?).

This code works for "small" `r.Body` but when the it's bigger than `3966` the `io.Copy` stops and the output form the print is `3966 context canceled`. So basically the https://github.com/go-kit/kit/blob/4914a6960de6e1a81504d2b8a9dd9f207d20c757/transport/http/client.go#L95 is called and the body lost.

Maybe I'm overcomplicating it using the `io.Pipe` but I thought it was the best solution for what I had in mind.

Maybe the `defer cancel()` should be called once the `r.Body.Close()` is called as it's I have the `BufferedStream` option no? IDK if that's even possible haha.

Thanks!

Ok, so I had some reading on the topic https://medium.com/the-software-firehose/canceling-requests-in-go-using-context-b4b28a118da8 and https://blog.cloudflare.com/the-complete-guide-to-golang-net-http-timeouts/ and I think that my assumptions where correct, if the `http.Request` is cancelled the `http.Response.Body` is closed too.

So then I tried to reproduce it on the `go-kit` test to make sure that It was, in fact, something not related to my code (which could be too :smile:) and I did achieve to reproduce it on the test for [`BufferedStream`](https://github.com/go-kit/kit/blob/master/transport/http/client_test.go#L100) basically adding a big `testbody` https://github.com/go-kit/kit/blob/master/transport/http/client_test.go#L102 (like `len(testbody) == 8400`) and a `time.Sleep(1 *time.Second)` (to simulate work here https://github.com/go-kit/kit/blob/master/transport/http/client_test.go#L133) the output was:
```
--- FAIL: TestHTTPClientBufferedStream (1.00s)
    client_test.go:139: want "EOF", have "context canceled"
FAIL
```
And some times (IDK why If I change the sleep time or something it switches from one to the other):
```
--- FAIL: TestHTTPClientBufferedStream (1.00s)
    client_test.go:144: want "EOF", have "read tcp 127.0.0.1:47002->127.0.0.1:34551: use of closed network connection"                                                                                             
FAIL
```

Removing the logic for cancelling (just for testing purposes) it returns another error:

```
--- FAIL: TestHTTPClientBufferedStream (1.00s)
    client_test.go:139: want "EOF", have %!q(<nil>)
FAIL
```

Which makes sense with the `io.Reader` [documentation](https://golang.org/pkg/io/#Reader):

> [...] An instance of this general case is that a Reader returning a non-zero number of bytes at the end of the input stream may return either err == EOF or err == nil. The next Read should return 0, EOF. [...]


And if I repeat https://github.com/go-kit/kit/blob/master/transport/http/client_test.go#L135 again it works (the read bytes of the first one is all of them `8400 <nil>`, and the second on is `0 EOF` so as expected).

So now that I know that this is an error, I tried to find a solution for it. First of all I did try to find a way to know when the `Response.Body` was closed, so then we can call the `cancel()`, as with the `BufferedStream` we left the `Response.Body` open and it has to be closed by the user, or the context. So I find no way to do it and the the solution would be to wrap the `Body` with a custom `Close()` that when it's call it calls the `Request.Body.Close()` and also `cancel()`.

The idea would be something like this:

```go
type bodyWithCancel struct {
        io.ReadCloser

        cancel context.CancelFunc
}

func (bwc bodyWithCancel) Close() error {
        bwc.ReadCloser.Close()
        bwc.cancel()
        return nil
}
```

In terms of _use_ this would mean that when the `BufferedStream` is used, the user should call `Close()` to the Body, which should not be that bad as with this option you are taking care of the Body already.

In terms of _code_ it would mean that on the `Client.Endpoint` we would have to call `cancel()` manually on all the possible errors (which is not nice), and when `BufferedStream` is validated, if it's `true` then we'll replace the `Response.Body` by:

```go
if c.bufferedStream {
  resp.Body = bodyWithCancel{ReadCloser: resp.Body, cancel: cancel}
} else {
  defer resp.Body.Close()
}
```

I'm commenting all this, because IDK if it may have potential side effects that I'm not aware of, or if it's a good solution (or wanted solution). So I prefer to comment it before opening a PR. Maybe there is another solution for the problem but I did not see any other :cry:

--- a/transport/http/client.go
+++ b/transport/http/client.go
@@ -85,7 +85,7 @@ func ClientFinalizer(f ...ClientFinalizerFunc) ClientOption {
 
 // BufferedStream sets whether the Response.Body is left open, allowing it
 // to be read from later. Useful for transporting a file as a buffered stream.
-// That body has to be Closed to propery end the request
+// That body has to be Closed to propery end the request.
 func BufferedStream(buffered bool) ClientOption {
 	return func(c *Client) { c.bufferedStream = buffered }
 }

--- a/transport/http/client.go
+++ b/transport/http/client.go
@@ -85,6 +85,7 @@ func ClientFinalizer(f ...ClientFinalizerFunc) ClientOption {
 
 // BufferedStream sets whether the Response.Body is left open, allowing it
 // to be read from later. Useful for transporting a file as a buffered stream.
+// That body has to be Closed to propery end the request
 func BufferedStream(buffered bool) ClientOption {
 	return func(c *Client) { c.bufferedStream = buffered }
 }
@@ -132,6 +133,8 @@ func (c Client) Endpoint() endpoint.Endpoint {
 			return nil, err
 		}
 
+		// If we expect a buffered stream, we don't cancel the context when the endpoint returns.
+		// Instead, we should call the cancel func when closing the response body.
 		if c.bufferedStream {
 			resp.Body = bodyWithCancel{ReadCloser: resp.Body, cancel: cancel}
 		} else {

--- a/transport/http/client.go
+++ b/transport/http/client.go
@@ -5,6 +5,7 @@ import (
 	"context"
 	"encoding/json"
 	"encoding/xml"
+	"io"
 	"io/ioutil"
 	"net/http"
 	"net/url"
@@ -92,7 +93,6 @@ func BufferedStream(buffered bool) ClientOption {
 func (c Client) Endpoint() endpoint.Endpoint {
 	return func(ctx context.Context, request interface{}) (interface{}, error) {
 		ctx, cancel := context.WithCancel(ctx)
-		defer cancel()
 
 		var (
 			resp *http.Response
@@ -112,10 +112,12 @@ func (c Client) Endpoint() endpoint.Endpoint {
 
 		req, err := http.NewRequest(c.method, c.tgt.String(), nil)
 		if err != nil {
+			cancel()
 			return nil, err
 		}
 
 		if err = c.enc(ctx, req, request); err != nil {
+			cancel()
 			return nil, err
 		}
 
@@ -126,11 +128,15 @@ func (c Client) Endpoint() endpoint.Endpoint {
 		resp, err = c.client.Do(req.WithContext(ctx))
 
 		if err != nil {
+			cancel()
 			return nil, err
 		}
 
-		if !c.bufferedStream {
+		if c.bufferedStream {
+			resp.Body = bodyWithCancel{ReadCloser: resp.Body, cancel: cancel}
+		} else {
 			defer resp.Body.Close()
+			defer cancel()
 		}
 
 		for _, f := range c.after {
@@ -144,6 +150,20 @@ func (c Client) Endpoint() endpoint.Endpoint {
 
 		return response, nil
 	}
+}
+
+// bodyWithCancel is a wrapper for an io.ReadCloser with also a
+// cancel function which is called when the Close is used
+type bodyWithCancel struct {
+	io.ReadCloser
+
+	cancel context.CancelFunc
+}
+
+func (bwc bodyWithCancel) Close() error {
+	bwc.ReadCloser.Close()
+	bwc.cancel()
+	return nil
 }
 
 // ClientFinalizerFunc can be used to perform work at the end of a client HTTP

--- a/transport/http/client_test.go
+++ b/transport/http/client_test.go
@@ -99,7 +99,7 @@ func TestHTTPClient(t *testing.T) {
 
 func TestHTTPClientBufferedStream(t *testing.T) {
 	var (
-		testbody = "testbody"
+		testbody = string(make([]byte, 6000))
 		encode   = func(context.Context, *http.Request, interface{}) error { return nil }
 		decode   = func(_ context.Context, r *http.Response) (interface{}, error) {
 			return TestResponse{r.Body, ""}, nil
@@ -129,6 +129,7 @@ func TestHTTPClientBufferedStream(t *testing.T) {
 	if !ok {
 		t.Fatal("response should be TestResponse")
 	}
+	time.Sleep(time.Second * 1)
 
 	// Check that response body was NOT closed
 	b := make([]byte, len(testbody))

--- a/transport/http/client_test.go
+++ b/transport/http/client_test.go
@@ -98,8 +98,12 @@ func TestHTTPClient(t *testing.T) {
 }
 
 func TestHTTPClientBufferedStream(t *testing.T) {
+	// bodysize has a size big enought to make the resopnse.Body not an instant read
+	// so if the response is cancelled it wount be all readed and the test would fail
+	// The 6000 has not a particular meaning, it big enough to fulfill the usecase.
+	const bodysize = 6000
 	var (
-		testbody = string(make([]byte, 6000))
+		testbody = string(make([]byte, bodysize))
 		encode   = func(context.Context, *http.Request, interface{}) error { return nil }
 		decode   = func(_ context.Context, r *http.Response) (interface{}, error) {
 			return TestResponse{r.Body, ""}, nil
@@ -130,6 +134,7 @@ func TestHTTPClientBufferedStream(t *testing.T) {
 		t.Fatal("response should be TestResponse")
 	}
 	defer response.Body.Close()
+	// Faking work
 	time.Sleep(time.Second * 1)
 
 	// Check that response body was NOT closed

--- a/transport/http/client_test.go
+++ b/transport/http/client_test.go
@@ -129,6 +129,7 @@ func TestHTTPClientBufferedStream(t *testing.T) {
 	if !ok {
 		t.Fatal("response should be TestResponse")
 	}
+	defer response.Body.Close()
 	time.Sleep(time.Second * 1)
 
 	// Check that response body was NOT closed

